CIP-ledger-evolution/README.md

```diff
@@ -0,0 +1,302 @@
+
+The STS framework leaves both cryptographic primitives and the serialization format abstract.
+These specifications need to be complete enough to realize a full implementation of the ledger
+given the cryptographic primitives and serialization format.
+The cryptographic primitives are described as appendices to the STS specifications,
+and the serialization format is given as a
+[CDDL file](https://www.rfc-editor.org/rfc/rfc8610).
+There SHOULD be one STS specification per ledger era
+(the Allegra ledger era is currently combined with the Mary era).
+
+From the Byron to the Babbage ledger eras, the STS frameworks were written in $\LaTeX$.
+Starting in the Conway ledger era, literate Agda will be used.
+During the transition from $\LaTeX$ to literate Agda, we will take advantage
+of the ability to substitute $\LaTeX$ in place of Agda code when needed for expedience.
+With time, the Agda specification will gloriously blur the line between specification
+and reference implementation,
+just as the type system upon which Agda is built blurs the line between proof and program.
+
+### Ledger eras
+
+A ledger era is a collection of features added to the ledger which are introduced at a hard fork.
+The existing ledger eras, with a very simplistic description is given below.
+
+|name|new features|
+| --- | --- |
+|Byron|initial UTxO ledger|
+|Shelley|decentralized block production, stake delegation|
+|Allegra|timelock scripts|
+|Mary|multi-assets|
+|Alonzo|Plutus scripts|
+|Babbage|improved Plutus script contexts|
+|Conway|constitutional committee replacement, SPO voting on hard forks|
+
+
+### Ledger state
+
+The ledger state is the source of truth from which the network derives meaning.
+For performance reasons, we require it to be the minimal state needed
+to derive this value and validate blocks.
+
+There is currently one exception to the minimality, which will be deprecated in the days ahead,
+namely the stake pool ranking information which is used by Daedalus.
+
+#### Ledger events
+
+Some provenance about the ledger calculations is provided by ledger events.
+These events come with zero cost to a running node if not used and are not stored in the ledger state.
+
+### Soft forks and Hard forks
+
+We make the following definitions:
+
+**Hard fork** - A *hard fork* is a change to the protocol (not restricted to the ledger)
+resulting in a single new (CBOR) block being valid.
+
+**Soft fork** - A *soft fork* is a change to the protocol (not restricted to the ledger)
+resulting in fewer blocks being valid.
+A good way to think about soft forks is the property that a node which only validates
+(and does not produce blocks) does not need a software upgrade in order to continue following the block chain.
+Soft forks do not seem to come up often in practice.
+
+### Serialization
+
+Transactions and blocks are serialized with
+[CBOR](https://www.rfc-editor.org/rfc/rfc7049)
+and specified with
+[CDDL file](https://www.rfc-editor.org/rfc/rfc8610).
```

Comment:
With very few exceptions, no, the CDDL describes what we allow for _deserialization_ (the only exception I can think of right now is the custom binary address format, but that has lengthy comments in our CDDL spec). It's true that CBOR is not by default canonical, and it is true that our ledger _serializers_ have to make choices, but that is irrelevant. Folks do often get tripped up when it comes to hashing, they attempt to mimic _our_ serialization choices to re-create the hash. but this is a security problem in itself, no one should be re-serializing, folks need to save the serialized data (which the ledger goes to great lengths to do, and which we've written about in many places, I'm happy to point you to those public discussions if you are interested).

Reply:
OK, so the CDDL in itself is unambiguous and two implementations following the same CDDL should be compatible?

Reply:
correct!

Reply:
We probably should have (or already have?) compatibility tests somewhere :)

Reply:
of course we do! that's one of the first things we do when we kick start a new era, we make the new transaction body, write the serializers, and make sure they round trip and adhere to the CDDL! Here's the tests for the conway era: https://github.com/input-output-hk/cardano-ledger/blob/24d7a68b00c2e71c4ba9491bacb6be1970015602/eras/conway/test-suite/test/Test/Cardano/Ledger/Conway/Serialisation/CDDL.hs They use a ruby tool, but goodness it would be grand to have a Haskell library for it..

And behold the twiddler: https://github.com/input-output-hk/cardano-ledger/blob/24d7a68b00c2e71c4ba9491bacb6be1970015602/eras/conway/test-suite/test/Test/Cardano/Ledger/Conway/Serialisation/CDDL.hs The twiddling allows one to mutate an existing encoding into another isomorphic CBOR compliant version.

Reply:
Of course you do! I was thinking of a tool that could be used by wannabe implementors to test _their_ serialisation library/code

Reply:
ah! The only thing I am aware of is the ruby :gem::  https://rubygems.org/gems/cddl/versions/0.8.5

Reply:
and if they are using Haskell, they might find our twiddlers helpful.